internal/observability/middleware_test.go

```diff
@@ -947,60 +987,143 @@ func TestMiddlewareMetrics(t *testing.T) {
+			wantCalls:          1,
+			wantSuccesses:      0,
+			applicationErr:     true,
+			applicationErrName: "SomeError",
+			wantCallerFailures: map[failureTags]int{
+				failureTags{
+					errorTag:     "application_error",
+					errorNameTag: "SomeError",
+				}: 1,
+			},
+		},
+		{
+			desc:               "application error name with YARPC code - caller failure",
+			wantCalls:          1,
+			wantSuccesses:      0,
+			applicationErr:     true,
+			applicationErrName: "SomeError",
+			applicationErrCode: &yErrAlreadyExists,
+			wantCallerFailures: map[failureTags]int{
+				failureTags{
+					errorTag:     "already-exists",
+					errorNameTag: "SomeError",
+				}: 1,
+			},
+		},
+		{
+			desc:               "application error name with YARPC code - server failure",
+			wantCalls:          1,
+			wantSuccesses:      0,
+			applicationErr:     true,
+			applicationErrName: "InternalServerPain",
+			applicationErrCode: &yErrCodeUnknown,
+			wantServerFailures: map[failureTags]int{
+				failureTags{
+					errorTag:     "unknown",
+					errorNameTag: "InternalServerPain",
+				}: 1,
+			},
+		},
+		{
+			desc:               "application error with YARPC code and empty name",
+			wantCalls:          1,
+			wantSuccesses:      0,
+			applicationErr:     true,
+			applicationErrName: "",
+			applicationErrCode: &yErrAlreadyExists,
+			wantCallerFailures: map[failureTags]int{
+				failureTags{
+					errorTag:     "already-exists",
+					errorNameTag: _notSet,
+				}: 1,
 			},
 		},
 	}
 
 	newHandler := func(t test) fakeHandler {
-		return fakeHandler{err: t.err, applicationErr: t.applicationErr}
+		return fakeHandler{
+			err:                t.err,
+			applicationErr:     t.applicationErr,
+			applicationErrName: t.applicationErrName,
+			applicationErrCode: t.applicationErrCode,
+		}
 	}
 
 	newOutbound := func(t test) fakeOutbound {
-		return fakeOutbound{err: t.err, applicationErr: t.applicationErr}
+		return fakeOutbound{
+			err:                t.err,
+			applicationErr:     t.applicationErr,
+			applicationErrName: t.applicationErrName,
+			applicationErrCode: t.applicationErrCode,
+		}
 	}
 
 	for _, tt := range tests {
 		validate := func(mw *Middleware, direction string, rpcType transport.Type) {
 			key, free := getKey(req, direction, rpcType)
 			edge := mw.graph.getEdge(key)
 			free()
-			assert.Equal(t, int64(tt.wantCalls), edge.calls.Load())
-			assert.Equal(t, int64(tt.wantSuccesses), edge.successes.Load())
-			assert.Equal(t, int64(0), edge.panics.Load())
-			for tagName, val := range tt.wantCallerFailures {
-				assert.Equal(t, int64(val), edge.callerFailures.MustGet(_error, tagName).Load())
+			assert.EqualValues(t, tt.wantCalls, edge.calls.Load(), "expected calls mismatch")
+			assert.EqualValues(t, tt.wantSuccesses, edge.successes.Load(), "expected successes mismatch")
+			assert.EqualValues(t, 0, edge.panics.Load(), "expected panics mismatch")
+			for tagValues, val := range tt.wantCallerFailures {
```

Comment:
nit: seeing `tagValues` and `val` together isn't intuitive. Can we change this to `failureTags` and `num`?
```suggestion
			for failureTags, num := range tt.wantCallerFailures {
```

The tests in this package are already quite large so I'd like to be as explicit as we can.

Reply:
Sure! Will change it across the whole file.